### DISCUSSION TITLE
Channel order playing 5.1 audio files

### DISCUSSION THREAD

**Discussion by jcucurull** [MusicPlayerDaemon/MPD#2060] [Q&A] [answered]
Hello,

I have a Moodeaudio installed that uses MPD to reproduce audio files. However, when I try to reproduce multichannel FLACs the audio channels are messed up. Is there some way to configure the correct audio channel mapping for MPD? 

I have been doing some tests (https://moodeaudio.org/forum/showthread.php?tid=6519&pid=55404#pid55404) and when I reproduce the files from the command line tool "aplay" it only swaps the LFE and center channels, but when I use the MPD it swaps more channels. Can the channel order be configured at the MPD level?

Best regards,
Jordi.

**Comment by MaxKellermann** [comment]
There's the "route" filter plugin: https://mpd.readthedocs.io/en/stable/plugins.html#route
But this just sidesteps the problem. It would be better to fix the real cause of the swapped channels. Are your files okay? Are your speakers connected correctly? Does MPD misinterpret the channel order from the file?

**Reply by jcucurull** [reply]
Hello Max,

I am using a FLAC testing file that is exactly for the purpose of checking the correct mapping of the channels:

https://github.com/sfiera/flac-test-files/blob/master/surround51.flac

This file encodes the channels in this order: FL FR FC LFE RL RR . Also, using some multichannel music I have in FLAC I get the same behaviour than with the test file. 

If I try my speakers using the "speaker-test -c 6 -t wav" command, the one provided by ALSA I think, the test is successful. If I decode the FLAC file to WAV and I reproduce it with the Alsa play command "aplay" I get the FC and LFE channels switched. However, if I indicate by parameter the order in which they are encoded in the FLAC, it reproduces it flawlessly:

> flac -c -d surround51.flac | aplay -m FL,FR,FC,LFE,RL,RR

The fact that the speaker test works well and that indicating the channels order in the aplay command works also well, makes me think that there is no issue in the ALSA configuration regarding the output to the AVR (I use an AVR connected via HDMI). 

If I reproduce the FLAC or the WAV created from the FLAC in MPD, I got the following channel order (left column speakers, right column channel played):

Front left: Front left
Front right: Front right
Rear left: Front center
Rear right: LFE
Center: Rear right
LFE: Rear left

As you can see it is a different order than the one get when I reproduce with "aplay" without indicating the channel mapping. So it seems as if during the transmission of the audio stream between MPD and ALSA there were an incorrect reordering of the channels or incorrect channel mapping. 

That is why I was asking if there were some place in the configuration that would enable the possibility to define a channel mapping, for the output to ALSA or for the decoding of the FLAC files. Otherwise, do you have a suggestion of why this can be happening?

Thank you very much in advance.

**Reply by jcucurull** [reply]
Just an addition, I have tested the route filter workaround and it works well but just for 5.1 audio. After configured it forces 5.1 output in all the cases. 

This would be more or less fine, but there is some glitch that makes that stereo audio songs are played with different channel mappings each time a song is started. So first you hear it normally for the FL and FR speakers, but the next one is played through FC and LFE, later RL and RR, and so on.

The configuration applied is the following:

```
music_directory "/var/lib/mpd/music"
playlist_directory "/var/lib/mpd/playlists"
db_file "/var/lib/mpd/database"
log_file "/var/log/mpd/log"
pid_file "/var/run/mpd/pid"
state_file "/var/lib/mpd/state"
sticker_file "/var/lib/mpd/sticker.sql"
user "mpd"
group "audio"
bind_to_address "any"
port "6600"
log_level "default"
restore_paused "yes"
auto_update "no"
follow_outside_symlinks "yes"
follow_inside_symlinks "yes"
zeroconf_enabled "no"
zeroconf_name "Moode MPD"
filesystem_charset "UTF-8"
metadata_to_use "+comment"
replaygain "off"
replaygain_preamp "0"
volume_normalization "no"
audio_buffer_size "4096"
max_output_buffer_size "131072"
max_playlist_length "16384"
max_connections "128"

decoder {
plugin "ffmpeg"
enabled "yes"
}

input {
plugin "curl"
proxy ""
proxy_user ""
proxy_password ""
}

filter {
plugin "route"
name "fix51"
routes "0>0,1>1,2>5,3>4,4>2,5>3"
}

resampler {
plugin "soxr"
quality "high"
threads "1"
}

audio_output {
type "alsa"
name "ALSA Default"
device "_audioout"
mixer_type "none"
dop "no"
stop_dsd_silence "no"
thesycon_dsd_workaround "no"
filters "fix51"
}

audio_output {
type "alsa"
name "ALSA Bluetooth"
device "_audioout"
mixer_type "software"
}

audio_output {
type "httpd"
name "HTTP Server"
port "8000"
encoder "lame"
bitrate "320"
tags "yes"
always_on "yes"
format "44100:16:2"
}

```

Is there a manner to apply the route filter only on 5.1 audio, so it does not affect stereo audio?

I suspect that the channel switching in this case is due to a known bug in RPi (https://forums.raspberrypi.com/viewtopic.php?p=2194020#p2194020) that is triggered by the fact of reproducing stereo over 5.1 output.

**Reply by jcucurull** [reply]
After all the testing and checks I have one question. Is MPD using the channel mapping API (https://docs.kernel.org/sound/designs/channel-mapping-api.html) to inform about the channel mapping used to ALSA?

If yes, from where MPD is obtaining the mapping? From the FLAC files? As far as I know FLAC files have a fixed channel mapping that is the following for 5.1 audio: FL,FR,FC,LFE,RL,RR

In the tests above, it is clear that exchanging the tuples (FC,LFE) and (RL,RR) and also switching FC and LFE solves the problem. So it seems MPD is informing to ALSA of a channel mapping different from the standard one in FLAC files (https://xiph.org/flac/format.html#frame_header).

**Reply by MaxKellermann** [reply]
Internally, MPD uses the FLAC convention for channel assignment: https://xiph.org/flac/format.html
ALSA uses a slightly different channel assignment; that's why MPD's ALSA output plugin has to swap some channels; `git grep` keyword is "alsa_channel_order".
MPD does not use any ALSA channel mapping API. I do not know this API; what you posted is the kernel/userspace ABI description, but MPD doesn't use that, but uses libasound instead.

**Reply by MaxKellermann** [reply]
> The fact that the speaker test works well and that indicating the channels order in the aplay command works also well, makes me think that there is no issue in the ALSA configuration

If you explicitly need to ask aplay to change the order of channels, then this indicates the exact opposite: that something in your setup is wrong, and MPD may be working correctly, after all. As noted in the other thread you created in this discussion, the comment at the top of the file explains the FLAC channel ordering (apparently correct in your file) and the default ALSA channel ordering (which your setup deviates from).

**Comment by jcucurull** [comment]
Where should I do the grep? In the conf files? In the source code?

El dv., 21 de juny 2024, 11:04, Max Kellermann ***@***.***>
va escriure:

**Reply by MaxKellermann** [reply]
In the MPD git repository. That shows you where the code is that does this.

**Reply by jcucurull** [reply]
I have seen the place where you swap the channels: 

https://github.com/MusicPlayerDaemon/MPD/blob/381215fd733c3f9a6bd34dc16191434e0abb1b77/src/pcm/Order.cxx#L62

What I understand here is that you swap the last two tuples of channels, i.e. (FC,LFE)<->(RL,RR). From my tests with "aplay" the only change I had to apply was swapping FC with LFE, nothing else. Instead, using MPD I had to apply a swapping with the routing plugin that is exactly the opposite you are doing in the code referred above, so I was undoing your swapping.

Are you sure this swapping you do is needed?  Could this order be different depending on the hardware? (in my case I use HDMI in RPi 3b+)

About the Channel Order API that I was mentioning, it is not only at kernel level, look for example at the call to set the mapping from the alsa lib API: https://www.alsa-project.org/alsa-doc/alsa-lib/group___p_c_m.html#ga60ee7d2c2555e21dbc844a1b73839085

**Reply by MaxKellermann** [reply]
Did you see the comment at the top of the file?

**Reply by MaxKellermann** [reply]
I think implementing the libasound chmap API is the right way to solve your problem. I suggest you write a feature request for this.

**Reply by jcucurull** [reply]
I have been checking the comment and also the "surround51.conf" and "surround71.conf" in my system. As you mention, the default channel mapping seems aligned with your swapping of channels. However, I am not sure if these profiles are used when the audio stream is sent via HDMI, e.g. there is a "hdmi.conf" in the same folder. I am researching on it.

**Reply by jcucurull** [reply, marked as answer]
I have managed to solve the issue applying a remaping at the level of ALSA. To be more specific, I have created a custom device that is a plugin that remaps the channels including this in "/etc/asound.conf":

```
pcm.!hdmi-remap {
  type plug
  slave.pcm "_audioout"
  ttable {
	   0.0= 1
	   1.1= 1
	   2.4= 1
	   3.5= 1
	   4.3= 1
	   5.2= 1
	   6.6= 1
	   7.7= 1
  }
}
```

This works better than the approach of the route plugin of MPD because it does not force the stereo audio to be played as multichannel (stereo is still played in two channels). 

However, as Max mentioned the way to go is to implement in MPD the libasound chmap API. Thus by now I will use this workaround, but I will also add a feature request to use the mentioned API. 

Thank you for all the feedback and help provided!!!
